### ISSUE TITLE
Feature request: Allow setting document language for hyphenation

### ISSUE THREAD

**Issue by Maia-Everett** [naturalcrit/homebrewery#1343]
It would be nice if Homebrewery allowed setting document language other than English. All it would do would be placing a `lang` attribute on the appropriate HTML element.

The main use case is hyphenation, which requires the correct HTML `lang` attribute to be set. Specifically: I would like to typeset documents in English and Russian with hyphenation. While I can set `hyphens: auto` in CSS, currently neither language works because there is no `lang` attribute in the document, so the browser doesn't know which language to hyphenate for.

**Comment by ericscheid** [comment]
The appropriate element would likely be the `<div class="pages">`, which wraps all the individual pages.

In the brew properties panel we could put a simple text field with default value of `en`. We'd need to allow for more than 2 characters as valid lang attribute values includes "zh-Hans", "zh-Hant", "ru-Cyrl-BY" etc.

Hmm .. the max length can be figured from https://www.ietf.org/rfc/bcp/bcp47.txt .. it's late, I'm tired, and I left my ABNF decoder ring at home.

**Comment by G-Ambatte** [comment]
> The appropriate element would likely be the `<div class="pages">`, which wraps all the individual pages.

From my (admittedly brief) research into the subject, it seems like this could be fixed with a snippet like:

```
<style>
  .pages {
    lang: "en";
  }
</style>
```

Although `.phb#pN { lang:"en" }`  (where `N` is the desired page number) might also be useful, for more targeted language selection.

**Comment by ericscheid** [comment]
Alas, no. 

You can target elements with lang with CSS 
e.g. ` .phb :lang(it) { color: green; } `

You can set the lang of elements in the brew, but you'd need to write the html.
e.g. `<p lang="es">Hola!</p>`

But you can't set the lang of an element with CSS.

You can test that last statement with:
```
<style>
  .pages { lang: "es"; }
  .phb :lang(es) { background-color: yellow; }
</style>

## Hello!

<h2 lang="fr">Bonjour!</h2>

<h2 lang="es">Hola!</h2>
```

**Comment by G-Ambatte** [comment]
Aha! The old "solution is simple, elegant, easily implemented, and only slightly hampered by the fact that it does not solve the problem".

I suspect we modify the `template.js` file to incorporate `<html lang={$this.state.brewLang}>`; combined with a way to set it in the Brew Metadata (defaulting to `en`), might actually work.

**Comment by calculuschild** [comment]
Yep this can be a metadata option. That's probably the cleanest way to do it.

**Comment by Trokkin** [comment]
Hyphenation does not work even for english text w/ firefox by default, it would if I add `lang=""` to `<body>` by hand with developer tools. +1 to this issue.

**Comment by ericscheid** [comment]
> In the brew properties panel we could put a simple text field with default value of en. We'd need to allow for more than 2 characters as valid lang attribute values includes "zh-Hans", "zh-Hant", "ru-Cyrl-BY" etc.

We should also guard against XSS by forcefully zapping all non alphanum chars with `-`.

**Comment by Gazook89** [comment]
> Yep this can be a metadata option. 

> The appropriate element would likely be the <div class="pages">

>  suspect we modify the template.js file to incorporate <html lang={$this.state.brewLang}>

Just want to clarify-- changing the language in the properties/metadata editor should apply a `lang` attribute to `.pages` inside the iframe, correct?    And thus, *should not* impact template.js?   Changing the template.js would change the lang attribute for the whole site, UI included, which I don't think we want to do...or at least, not by using the metadata editor on each brew.   I could be wrong about what template.js does, so let me know if I got that wrong.

---

In a branch I have a **lang** input field in the properties editor, and it saves the input to the brew.  However, I can't figure out how to pass that brew property to the iframe to insert into `.pages`....?  If anyone has helpful hints, let me know.

**Comment by Gazook89** [comment]
Should this be a regular free-form text input, or a dropdown of choices?  Obviously there are a lot of possible choices especially if going into scripts and regions (in addition to language), so i'm not sure that is super practical without setting some arbitrary limit on what is included.  

Right now i'm leaning towards text input, with `Intl.DisplayNames()` ([here on MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DisplayNames/DisplayNames)) on the field to give a readable form of the code once it is entered.   I'm not sure how much I can do for text validation, though, without using a 3rd party library.

**Comment by ericscheid** [comment]
> Should this be a regular free-form text input, or a dropdown of choices?

A data list of likely choices on a free-form text input.

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/datalist

```
<label for="lang">Languages: </label>
<input type="text" name="lang" id="lang" list="langs" autocomplete="off">
<datalist id="langs">
  <option value="en">English</option>
  <option value="en-GB">English (Great Britain)</option>
  <option value="en-AU">English (Australia)</option>
  <option value="en-US">English (USA)</option>
  <option value="fr">Français </option>
  <option value="fr-CA">Français (Canadian)</option>
  <option value="de">Deutsch</option>
  <option value="de-CH">Schwiizerdütsch</option>
  <option value="dk">Dansk</option>
  <option value="jp-JP">日本語</option>
</datalist>
```

<img width="334" alt="demo of the above code" src="https://user-images.githubusercontent.com/2605467/197316968-38528e8f-df98-4cab-8255-6146c94ca71d.png">


NB:
1. <datalist> supported by all modern browsers, including accessibility concerns
2. <datalist> allows selection of option, or manual entry of custom values
3. language lists should be in the referenced language to aid the user in selecting *their* language
4. the actual list of languages presented above are *exampli gratia* for code purposes only
5. good practice to turn off the auto-complete on the input — various browsers will behave weirdly UI-wise
6. probably should have some minimal form of validation such as `[a-z]{2,3}(-.*)?` to avoid entry of common language _names_ e.g. "English"

https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/lang

**Comment by Gazook89** [comment]
So I have the above set up except instead of hardcoding both the code and the 'name' it only needs the code, and the display name is automatically given in that language.  Same end result, but with less work.

I am working on setting the regexp pattern now.  I see your basic pattern `[a-z]{2,3}(-.*)?` and that seems good enough to prevent entering 'English' or similar names...is that 'good enough' for our purposes?  Or should we use a [more robust pattern as given here](https://stackoverflow.com/a/60899733):

```regexp
^(((en-GB-oed|i-ami|i-bnn|i-default|i-enochian|i-hak|i-klingon|i-lux|i-mingo|i-navajo|i-pwn|i-tao|i-tay|i-tsu|sgn-BE-FR|sgn-BE-NL|sgn-CH-DE)|(art-lojban|cel-gaulish|no-bok|no-nyn|zh-guoyu|zh-hakka|zh-min|zh-min-nan|zh-xiang))|((([A-Za-z]{2,3}(-([A-Za-z]{3}(-[A-Za-z]{3}){0,2}))?)|[A-Za-z]{4}|[A-Za-z]{5,8})(-([A-Za-z]{4}))?(-([A-Za-z]{2}|[0-9]{3}))?(-([A-Za-z0-9]{5,8}|[0-9][A-Za-z0-9]{3}))*(-([0-9A-WY-Za-wy-z](-[A-Za-z0-9]{2,8})+))*(-(x(-[A-Za-z0-9]{1,8})+))?)|(x(-[A-Za-z0-9]{1,8})+))$
```

Note, on initial testing with this, it still allows for "English" as a valid entry, so i'm not sure if this regexp needs tweaking or if that is desirable.   

This also only is a regexp pattern....it would allow for invalid values as long as they fit the pattern.  So instead, we could use a library like [IETF Language Tags](https://github.com/sozialhelden/ietf-language-tags) which would provide methods for checking against the actual list of language codes.

I'm just not sure how deep we want to go?   If we settle for just a limited list, it isn't hard to add languages later....it would just be adding the correct code to an array without even worrying about the displayName, since that is automatically taken care of.

I'll upload my branch today.

**Comment by Gazook89** [comment]
Things left to do:

- [ ] currently the validation pattern just shows an error, but still accepts the value (and saves it to the document).  Likely it should only accept valid patterns.
- [ ] correctly set the default to 'en'
- [ ] confirm choices for the datalist.
